### DISCUSSION TITLE
Thanos store expensive blocks syncing

### DISCUSSION THREAD

**Discussion by tsroka** [thanos-io/thanos#5578] [General]
Hi,

In my setup I see excessive calls to azure's blob storage GetProperties API call (70% of all API calls). After a bit of digging I found out this happens when Thanos store periodically syncs blocks. More or less this works like this:

1. Each sync interval store wants to [sync blocks from object store](https://github.com/thanos-io/thanos/blob/main/cmd/thanos/store.go#L374)
2. Each sync calls [BaseFetcher.fetchMetadata](https://github.com/thanos-io/thanos/blob/main/pkg/block/fetcher.go#L307) and this method iterates through bucket and for each block directory calls [BaseFetcher.loadMeta](https://github.com/thanos-io/thanos/blob/main/pkg/block/fetcher.go#L223) method.
3. First thing `loadMeta` does, it checks for the existence of  a meta file in the directory, which translates to the GetProperties azure call (this is done regardless of the meta file being previously downloaded and cached).

I think this call may be unnecessary, but maybe I am missing something.  Because `loadMeta` caches the meta once it is loaded  for the first time (in memory and optionally on disk), removing the meta file existence check will  make syncing more reliable (less network calls) and save some $$$; I suspect that this check was added probably to handle the scenario where the block was deleted (there is not atomic deletion of the whole block directory, so first meta file is removed, then the rest), but given that the whole block directory will get deleted eventually (so it will not be listed when iterating through block directories as described in the second point) I do not see the reason for this check. 

Does anybody know why this check is necessary?

NOTE: this is especially annoying when using thanos store with sharding, since meta data is loaded for all blocks first, not only for blocks that are supposed to be served by a given shard (but this is a separate issue). 

Thanks,
Tomek.

**Comment by BooNny95** [comment]
Similar situation is for deletion-mark.json too... In our case it makes a lot of 404 responses...

I think the way to somehow reduce that syncs is to lower sync-block-duration time. Default is 3min...
